### PR TITLE
Migrate to self-descriptive substrate v14 metadata

### PR DESCRIPTION
**Pull request by breathx**
# Substrate updates
The long awaited pull request https://github.com/paritytech/substrate/pull/8615 is finally finished and merged. The changes made are extremely important and require updating our existing runtime and node base.
- Make migration to v14 metadata
- Update codebase according other latest changes
- Implement required `TypeInfo` for all custom structs

# Gear metadata interacting updates
Further more, we migrate our metadata collection algorithm on the `scale-info` and `polkadot-js-api` side (https://github.com/gear-tech/gear-js/pull/29) using `Registry` types, which now allows us support any Rust object types which `scale-info` supports.

Moreover, this PR reduces code base for getting metadata and adds more variety of declaring I/O types in contracts.
### Before
```rust
use gstd_meta::*;

meta! {
    title: "Title",
    init_input: MessageInitIn,
    init_output: MessageInitOut,
    input: MessageIn,
    output: MessageOut,
    extra: SubTypeOne, SubTypeTwo
}
```
### Now
```rust
gstd::metadata!{
    title: "Title",
    init:
        input: MessageInitIn,
        output: MessageInitOut,
    handle:
        input: MessageIn,
        output: MessageOut
}
// Subtypes includes automatically

// The macro requires title declaration,
// comma at the end of type lines (except the last one)
// and at least one type declaration in one of the sections: init, handle

// Another example
gstd::metadata!{
    title: "Title",
    init:
        input: MessageInitIn,
    handle:
        output: MessageOut
}

// And another one
gstd::metadata!{
    title: "Title",
    init:
        input: MessageInitIn
}
```

@gear-tech/dev

### CODE REVIEW

File: utils/wasm-proc/metadata-js/index.js [hunk 1]
```diff
@@ -31,8 +31,8 @@ exports.getWasmMetadata = async (wasmBytes) => {
     metadata.init_output = readMeta(memory, module.instance.exports.meta_init_output());
     metadata.input = readMeta(memory, module.instance.exports.meta_input());
     metadata.output = readMeta(memory, module.instance.exports.meta_output());
+    metadata.registry = readMeta(memory, module.instance.exports.meta_registry());
```
Comment (osipov-mit) [inline utils/wasm-proc/metadata-js/index.js:34]: meta_registry returns hex without prefix 0x? 
Maybe add it here? Like this:
```javascript
`0x${readMeta(memory, module.instance.exports.meta_registry)}`
```

File: utils/wasm-proc/metadata-js/test.js [hunk 2]
```diff
@@ -3,48 +3,26 @@ const assert = require('assert');
 const fs = require("fs");
 const path = require("path");
 const wasmMetadata = require(".");
+const api = require('@polkadot/types');
```
Comment (osipov-mit) [inline utils/wasm-proc/metadata-js/test.js:6]: You can make import like this 
```javascript
const { TypeRegistry } = require('@polkadot/types')
```

File: utils/wasm-proc/metadata-js/test.js [hunk 3]
```diff
@@ -3,48 +3,26 @@ const assert = require('assert');
...
 
 let wasmBytes = fs.readFileSync(
     path.join(__dirname, "../../../examples/target/wasm32-unknown-unknown/release/demo_meta.meta.wasm")
 );
 
 wasmMetadata.getWasmMetadata(wasmBytes).then(metadata => {
+    let regSource = "0x" + metadata.registry;
+    const reg = new api.TypeRegistry();
```
Comment (osipov-mit) [inline utils/wasm-proc/metadata-js/test.js:14]: And remove `api.` from here
Just 
```javascript
const reg = new TypeRegistry()
```

**Review by osipov-mit — CHANGES_REQUESTED** [review]

**Review by osipov-mit — APPROVED** [review]

File: gtest/src/js/mod.rs [hunk 4]
```diff
@@ -168,20 +168,20 @@ mod tests {
...
+            .convert(
+                "examples/target/wasm32-unknown-unknown/release/demo_meta.meta.wasm",
                 &MetaType::Input,
             )
-            .expect("Could not find file "),
-        );
+            .or(json.convert(
+                "target/wasm32-unknown-unknown/release/demo_meta.meta.wasm",
+                &MetaType::Input,
+            ));
```
Comment (shamilsan) [inline gtest/src/js/mod.rs:180]: It is not recommended to call expensive functions in the `.or()` method. Use the lazy alternative.
```suggestion
            .or_else(|_| json.convert(
                "target/wasm32-unknown-unknown/release/demo_meta.meta.wasm",
                &MetaType::Input,
            ));
```

**Comment by shamilsan** [comment]
I have a problem with building WASM too.
```
gear on  dn-migrate-to-substrate-metadata [$!] via 🦀 v1.55.0 took 36s 
❯ ./scripts/build-wasm.sh 
    Finished release [optimized] target(s) in 0.01s
    Finished release [optimized] target(s) in 0.01s
    Finished release [optimized] target(s) in 0.03s
    Finished release [optimized] target(s) in 0.04s
    Finished release [optimized] target(s) in 0.03s
    Finished release [optimized] target(s) in 0.03s
    Finished release [optimized] target(s) in 0.03s
    Finished release [optimized] target(s) in 0.04s
    Finished release [optimized] target(s) in 0.03s
    Finished release [optimized] target(s) in 0.04s
    Finished release [optimized] target(s) in 0.01s
    Finished release [optimized] target(s) in 0.01s
    Finished release [optimized] target(s) in 0.01s
    Finished release [optimized] target(s) in 0.01s
    Finished release [optimized] target(s) in 0.03s
    Finished release [optimized] target(s) in 0.04s
    Finished release [optimized] target(s) in 0.04s
    Finished release [optimized] target(s) in 0.04s
./scripts/build-wasm.sh: 27: /home/shamil/devel/gear-tech/gear/examples/../target/release/wasm-proc: not found
```

**Comment by breathx** [comment]
@shamilsan You need to build 'wasm-proc' util. I don't actually think that it should be included in 'build-wasm' script.
Run 'cargo build -p wasm-proc --release' and try again

**Comment by shamilsan** [comment]
> @shamilsan You need to build 'wasm-proc' util. I don't actually think that it should be included in 'build-wasm' script.
> Run 'cargo build -p wasm-proc --release' and try again

I'd add `cargo build -p wasm-proc --release` to the script as it should be self-sufficient.

File: gtest/src/js/mod.rs (hunk 4 above)
Comment (breathx) [reply gtest/src/js/mod.rs]: Done

**Review by ekovalev — APPROVED** [review]

**Review by shamilsan — APPROVED** [review]

**Review by ukint-vs — APPROVED** [review]

**Review by NikVolf — APPROVED** [review]

**Comment by NikVolf** [comment]
Great stuff 👍

**Pull request merged**